### PR TITLE
Draft: Make ShapeString editable by double-click - fixes #5885

### LINKED ISSUE #5885: ShapeString is not editable
Issue imported from https://tracker.freecad.org/view.php?id=3853

* **Reporter:** uwestoehr
* **Date submitted:** 2/22/2019
* **FreeCAD version:** 0.18
* **Category:** Bug
* **Status:** assigned
* **Tags:** Task View

# Original report text

- created with the Draft WB a ShapeString.

This opens a task dialog where you can specify settings

- after the creation, double-click on the ShapeString in the model tree

result: the ShapeString gets a yellow background but the task dialog doesn't show up. So FC is in editing mode but there is nothing to edit.
In the report panel I see
"This object type is not editable"

Since there exists a nice task dialog for ShapeStrings, it should be shown.

# Additional information

OS: Windows 7
Word size of OS: 64-bit
forum thread: https://forum.freecadweb.org/viewtopic.php?f=3&t=34370

Word size of FreeCAD: 64-bit
Version: 0.18.15959 (Git)
Build type: Release
Branch: master
Hash: 397418078a6f61e8c39cedfe1160adc2abd73510
Python version: 3.6.6
Qt version: 5.6.2
Coin version: 4.0.0a
OCC version: 7.3.0
Locale: German/Germany (de_DE)

# Other bug information

* **Priority:** normal
* **Severity:** minor
* **Category:** Bug
* **Updated:** 2/6/2021

# Discussion from Mantis ticket

### Comment by yorik 2019-02-22 13:27
Indeed that object has its edit mode not fully implemented. I'll do that after the 0.18 release

---

Did you find time to have a look meanwhile?

---

### Comment by Kunda1 2020-12-03 23:50
yorik should we bump this to 0.20 ?

---

### Comment by yorik 2020-12-16 11:19
No, let  me try to fix it before 0.19 ;)

### PR DESCRIPTION
**Pull request by marioalexis84**
With this changes, Draft ShapeString objects are editable when double-click on the model tree.
Two specific classes are added (ShapeStringTaskPanelCmd and ShapeStringTaskPanelEdit) to create the task panel used by the Draft_ShapeString command and by the view provider when the object is edited.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [x]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6721">pipeline status</a> for feature branch [PR_6721](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6721). Pipeline [511033501](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/511033501) was triggered at [504a45e](https://github.com/FreeCAD/FreeCAD/pull/6721/commits/504a45e0637b54bf0ea64d971146dda053b5a443). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Draft/draftviewproviders/view_shapestring.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#*   License along with this program; if not, write to the Free Software   *
+#*   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  *
+#*   USA                                                                   *
+#*                                                                         *
+#***************************************************************************
+
+
+"""Provides the viewprovider code for the Shapestring object."""
+
+import FreeCAD as App
```
Comment (benj5378) [inline src/Mod/Draft/draftviewproviders/view_shapestring.py:26]: Is `App` used?

**Comment by donovaly** [comment]
Many thanks. The PR fixes the issue for me.

However, I don't see the effect of changes. I have to close the dialog first. Ideally, when  change the string if the font side, I would like to see the change immediately.
Of course this wish is not important for the PR to be merged since it fixes the initial bug report in every case.

File: src/Mod/Draft/draftviewproviders/view_shapestring.py (hunk 1 above)
Comment (marioalexis84) [reply src/Mod/Draft/draftviewproviders/view_shapestring.py:26]: No, it is not used. Thanks for the review.

**Comment by marioalexis84** [comment]
I agree that ideally the update should be done when some property is changed. But at the moment the `onChanged` method of the Draft objects is not fully implemented.

**Comment by yorikvanhavre** [comment]
Although this is more a feature than a bugfix, I think it would be good to have for 0.20... Would you mid doing the fix mentioned by @benj5378 before?

**Comment by marioalexis84** [comment]
> Would you mid doing the fix mentioned by @benj5378 before?

Done. A new commit just to remove another unused import in the facebinder viewprovider.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6721">pipeline status</a> for feature branch [PR_6721](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6721). Pipeline [512426461](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/512426461) was triggered at [ed1d0e4](https://github.com/FreeCAD/FreeCAD/pull/6721/commits/ed1d0e4152d527e1b2de1c2ff6f817caa462d380). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by donovaly** [comment]
> Although this is more a feature than a bugfix,

For me it fixes a real bug since as normal user you are confronted with a message that there is an open task panel but there is none and you are somehow lost.
Moreover, since one cannot select a font file for ShapeStrings on Windows from within the properties editor, see https://forum.freecadweb.org/viewtopic.php?p=587201#p587201 and Roy's reply, we need a task panel to edit existing ShapeStrings.

Since Mario improved the PR as suggested, I'll merge it.

**Comment by marioalexis84** [comment]
@donovaly since you are using Windows, can you check if the issue reported [on the forum](https://forum.freecadweb.org/viewtopic.php?f=10&t=67869&sid=9d5668794d7238952d506836107b4ab2) still exist? It may be necessary to restore the type of dialog after edit the shapestring.

**Comment by donovaly** [comment]
> can you check if the issue reported [on the forum](https://forum.freecadweb.org/viewtopic.php?f=10&t=67869&sid=9d5668794d7238952d506836107b4ab2) still exist?

Sure, because I that was me who reported it ;-)

**Pull request merged**